### PR TITLE
core(pwa): revamp and move short_name_length audit

### LINKED ISSUE #348: manifest's "short_name" isn't required if "name" is short enough
See this thread for details: https://twitter.com/nekrtemplar/status/734483891071922182

Not sure about what is exactly "short enough", but 4 symbols seems to be.

### LINKED ISSUE #2698: short_name shouldn't be required
https://github.com/GoogleChrome/lighthouse/blob/master/lighthouse-core/audits/webapp-install-banner.js#L49

If the `name` property is short enough, you don't need `short_name`. The test should reflect this.

This is either a regression of https://github.com/GoogleChrome/lighthouse/issues/348, or wasn't fixed.

### PR DESCRIPTION
**Pull request by paulirish**
Two separate, but related changes in here:

1. webapp-install-banner audit doesn't check for a defined `short_name` any more. It's never technically been required, so we shouldn't assert it.
1. Just straight up remove short-name-length. The length of 12 characters was picked out of the air. And having a whole audit to verify `str.length <= 12` is kinda annoying. 
   * If we didn't remove it, I'd move it to PWA, set weight to 0, and flip it to `notApplicable: true` if no manifest is present. But honestly I don't think it's worth keeping it around. Open to arguments tho.

---------

Fixes #2698. See also https://github.com/GoogleChrome/lighthouse/issues/23#issuecomment-270461938
Fixes https://github.com/GoogleChrome/lighthouse/issues/686#issuecomment-339308030
Fixes #4860

Doesn't fix #348 because it makes a different conclusion.

### CODE REVIEW

**Comment by paulirish** [comment]
Welcome back!

* Now it's in PWA section
* Weight of 0 always, since it isn't included in PWA checklist requirements
* Failure of no manifest is communicated

### screenshot where there's no manifest at all

image

### screenshot where there's a failing manifest

image

**Comment by patrickhulce** [comment]
got some lint failures though

**Comment by paulirish** [comment]
> aside: is it just me or could multicheckaudit and byteefficiencyaudit use a bit of a refresh cleanup post I/O :)

yUPPPPPPPPP

File: lighthouse-core/audits/manifest-short-name-length.js [hunk 1]
```diff
@@ -28,26 +27,24 @@ class ManifestShortNameLength extends Audit {
...
-      const hasShortName = manifestValues.allChecks.find(i => i.id === 'hasShortName').passing;
-      if (!hasShortName) {
-        return {
-          rawValue: false,
-          debugString: 'No short_name found in manifest.',
-        };
-      }
+      const shortNameCheckIds = ['hasShortName', 'shortNameLength'];
+      manifestValues.allChecks.filter(item => shortNameCheckIds.includes(item.id)).forEach(item => {
+        if (!item.passing) failures.push(item.failureText);
```
Comment (brendankenny) [inline lighthouse-core/audits/manifest-short-name-length.js]: won't this have the other problem where if you don't have a short name you'll get both ``'Manifest does not have `short_name`'`` and ``'Manifest `short_name` will be truncated when displayed on the homescreen'``

**Comment by glittle** [comment]
As for the limit of 12, does this consider that Android word-wraps the text and shows on two lines?  So, two words, both 12 long, would be fine.

**Comment by paulirish** [comment]
> does this consider that Android word-wraps the text and shows on two lines?

It doesn't, but I don't see this on my phone. Android 8.1.0 on a Pixel 2. I've never seen the text below an icon get word-wrapping. :/

**Comment by glittle** [comment]
Hmm... it looks like on the desktop it does not, but in a folder it does!  I keep almost everything in folders, so didn't notice. (Sorry these screen captures are so big!)

screenshot_20180426-134730
screenshot_20180426-134953

File: lighthouse-core/gather/computed/manifest-values.js [hunk 2]
```diff
@@ -64,8 +64,10 @@ class ManifestValues extends ComputedArtifact {
       },
       {
         id: 'shortNameLength',
-        failureText: 'Manifest `short_name` will be truncated when displayed on the homescreen',
-        validate: manifest => !!manifest.value.short_name.value &&
+        failureText: `Manifest's \`short_name\` is too long (>${SUGGESTED_SHORTNAME_LENGTH} ` +
+          `characters) to be displayed on a homescreen without truncation`,
+        // Pass if there's no short_name. Don't want to report a non-existent string is too long
```
Comment (JoelEinbinder) [inline lighthouse-core/gather/computed/manifest-values.js]: I think that this should fail if there is no short_name and the name is longer than SUGGESTED_SHORTNAME_LENGTH. Or another audit like `hasSomeShortName` should fail. I wouldn't know that `short_name` exists without this audit.

**Comment by paulirish** [comment]
@patrickhulce @brendankenny ptal

**Comment by brendankenny** [comment]
should update the PR description (looks like you ended up doing the opposite)

File: lighthouse-core/audits/manifest-short-name-length.js [hunk 3]
```diff
@@ -26,28 +25,37 @@ class ManifestShortNameLength extends Audit {
...
+   * @return {Promise<{failures: Array<string>, manifestValues: LH.Artifacts.ManifestValues, notApplicable?: boolean}>}
    */
-  static audit(artifacts) {
+  static audit_(artifacts) {
+    /** @type {Array<string>} */
+    const failures = [];
+    /** @type {Array<string>} */
+    const warnings = [];
+
     return artifacts.requestManifestValues(artifacts.Manifest).then(manifestValues => {
```
Comment (brendankenny) [inline lighthouse-core/audits/manifest-short-name-length.js]: await

File: lighthouse-core/audits/manifest-short-name-length.js [hunk 4]
```diff
@@ -26,28 +25,37 @@ class ManifestShortNameLength extends Audit {
...
-        };
+      if (shortNameLengthCheck && !shortNameLengthCheck.passing) {
+        failures.push(shortNameLengthCheck.failureText);
       }
 
-      const isShortEnough = manifestValues.allChecks.find(i => i.id === 'shortNameLength');
-      return {
-        rawValue: !!isShortEnough && isShortEnough.passing,
-      };
+      return result;
```
Comment (brendankenny) [inline lighthouse-core/audits/manifest-short-name-length.js]: there's no warnings and only one possible failure, so why not drop mulitcheckAudit and do like
```js
const manifestValues = await artifacts.requestManifestValues(artifacts.Manifest);
// If there's no valid manifest, this audit is not applicable
if (manifestValues.isParseFailure) {
  return {
    rawValue: true,
    notApplicable: true,
  };
}

const shortNameCheck = manifestValues.allChecks.find(i => i.id === 'hasShortName');
const shortNameLengthCheck = manifestValues.allChecks.find(i => i.id === 'shortNameLength');

// If there's no short_name present, this audit is not applicable
if (shortNameCheck && !shortNameCheck.passing) {
  return {
    rawValue: true,
    notApplicable: true,
  };
}

if (shortNameLengthCheck && !shortNameLengthCheck.passing) {
  return {
    rawValue: false,
    explanation: `Failure: ${shortNameLengthCheck.failureText}.`,
  };
}

return {
  rawValue: true,
};
```

File: lighthouse-core/test/audits/manifest-short-name-length-test.js [hunk 5]
```diff
@@ -30,29 +30,30 @@ describe('Manifest: short_name_length audit', () => {
     artifacts.Manifest = null;
 
     return ManifestShortNameLengthAudit.audit(artifacts).then(result => {
-      assert.strictEqual(result.rawValue, false);
-      assert.strictEqual(result.explanation, undefined);
```
Comment (brendankenny) [inline lighthouse-core/test/audits/manifest-short-name-length-test.js]: update test description (doesn't fail)

File: lighthouse-core/test/audits/manifest-short-name-length-test.js [hunk 6]
```diff
@@ -30,29 +30,30 @@ describe('Manifest: short_name_length audit', () => {
...
+      assert.strictEqual(result.notApplicable, true);
     });
   });
 
   it('fails when an empty manifest is present', () => {
     const artifacts = generateMockArtifacts();
     artifacts.Manifest = manifestParser('{}', EXAMPLE_MANIFEST_URL, EXAMPLE_DOC_URL);
     return ManifestShortNameLengthAudit.audit(artifacts).then(result => {
-      assert.equal(result.rawValue, false);
-      assert.equal(result.explanation, 'No short_name found in manifest.');
```
Comment (brendankenny) [inline lighthouse-core/test/audits/manifest-short-name-length-test.js]: sames

File: lighthouse-core/test/audits/manifest-short-name-length-test.js [hunk 7]
```diff
@@ -30,29 +30,30 @@ describe('Manifest: short_name_length audit', () => {
...
+      assert.strictEqual(result.rawValue, true);
+      assert.strictEqual(result.notApplicable, true);
     });
   });
 
-  it('fails when a manifest contains no short_name and too long name', () => {
```
Comment (brendankenny) [inline lighthouse-core/test/audits/manifest-short-name-length-test.js]: sames

File: lighthouse-core/test/audits/manifest-short-name-length-test.js [hunk 8]
```diff
@@ -66,6 +67,8 @@ describe('Manifest: short_name_length audit', () => {
     artifacts.Manifest = manifestParser(manifestSrc, EXAMPLE_MANIFEST_URL, EXAMPLE_DOC_URL);
     return ManifestShortNameLengthAudit.audit(artifacts).then(result => {
       assert.equal(result.rawValue, false);
+      assert.ok(result.explanation.includes('without truncation'), result.explanation);
+      assert.ok(!result.explanation.includes('Manifest does not have `short_name`'));
```
Comment (brendankenny) [inline lighthouse-core/test/audits/manifest-short-name-length-test.js]: delete this line as it's not a possibility any longer? Want to assert not not applicable, though?

**Comment by patrickhulce** [comment]
:shipit:

**Comment by noyainrain** [comment]
Thanks for fixing this @paulirish :blush: 

I know it's been over a year, just a quick question related to this change. From the description of the PR, the original goal was to remove the `hasShortName` check:

>     1. webapp-install-banner audit doesn't check for a defined `short_name` any more. It's never technically been required, so we shouldn't assert it.

If I understand the discussion correctly, it was then agreed to make it **informative** instead:

> I'd propose turning it informative and marking not applicable when no manifest rather than killing it off entirely.

But in the end, the check still is mandatory:

```js
      // Technically shortname isn't required (if name is defined):
      //   https://cs.chromium.org/chromium/src/chrome/browser/installable/installable_manager.cc?type=cs&q=IsManifestValidForWebApp+f:cc+-f:test&sq=package:chromium&l=473
      // Despite this, we think it's better to require it anyway.
      // short_name is preferred for the homescreen icon, but a longer name can be used in
      // the splash screen and app title. Given the different usecases, we'd like to make it clearer
      // that the developer has two possible strings to work with.
      'hasShortName',
```

Is there a particular reason for keeping it required in the `installable-manifest` audit, instead of making it informative somehow?

(Problem background: Most PWA directories use the result of lighthouse `pwa-installable` to inform the user if a PWA is installable. For a manifest with a short `name` and no (redundant) `shortname` this produces a false-negative.)

**Pull request merged**